tests/filesystem/UploadTest.php

```diff
@@ -84,15 +84,91 @@ public function testAllowedFilesize() {
 		);
 		
+		// test upload into default folder
+		$u1 = new Upload();
 		$v = new UploadTest_Validator();
+		
 		$v->setAllowedMaxFileSize(array('txt' => 10));
+		$u1->setValidator($v);
+		$result = $u1->load($tmpFile);
+		$this->assertFalse($result, 'Load failed because size was too big');
 		
-		// test upload into default folder
-		$u1 = new Upload();
+		$v->setAllowedMaxFileSize(array('[doc]' => 10));
+		$u1->setValidator($v);
+		$result = $u1->load($tmpFile);
+		$this->assertFalse($result, 'Load failed because size was too big');
+		
+		$v->setAllowedMaxFileSize(array('txt' => 200000));
+		$u1->setValidator($v);
+		$result = $u1->load($tmpFile);
+		$this->assertTrue($result, 'Load failed with setting max file size');
+		
+		// check max file size set by app category
+		$tmpFileName = 'UploadTest-testUpload.jpg';
+		$tmpFilePath = TEMP_FOLDER . '/' . $tmpFileName;
+		file_put_contents($tmpFilePath, $tmpFileContent . $tmpFileContent);
+		
+		$tmpFile = array(
+			'name' => $tmpFileName,
+			'type' => 'image/jpeg',
+			'size' => filesize($tmpFilePath),
+			'tmp_name' => $tmpFilePath,
+			'extension' => 'jpg',
+			'error' => UPLOAD_ERR_OK,
+		);
+		
+		$v->setAllowedMaxFileSize(array('[image]' => '40k'));
 		$u1->setValidator($v);
 		$result = $u1->load($tmpFile);
+		$this->assertTrue($result, 'Load failed with setting max file size');
 		
+		$v->setAllowedMaxFileSize(array('[image]' => '1k'));
+		$u1->setValidator($v);
+		$result = $u1->load($tmpFile);
 		$this->assertFalse($result, 'Load failed because size was too big');
+		
+		$v->setAllowedMaxFileSize(array('[image]' => 1000));
+		$u1->setValidator($v);
+		$result = $u1->load($tmpFile);
+		$this->assertFalse($result, 'Load failed because size was too big');
+	}
+	
+	public function testGetAllowedMaxFileSize() {
+		// Check the max file size uses the config values
+		$configMaxFileSizes = array(
+			'[image]' => '1k',
+			'txt' => 1000
+		);
+		Config::inst()->update('Upload_Validator', 'default_max_file_size', $configMaxFileSizes);
+		$v = new UploadTest_Validator();
+		
+		$retrievedSize = $v->getAllowedMaxFileSize('[image]');
+		$this->assertEquals(1024, $retrievedSize, 'Max file size check on default values failed');
+		
+		$retrievedSize = $v->getAllowedMaxFileSize('txt');
+		$this->assertEquals(1000, $retrievedSize, 'Max file size check on default values failed');
+		
+		// Check instance values for max file size
+		$maxFileSizes = array(
+			'[doc]' => 2000,
+			'txt' => '4k'
+		);
+		$v = new UploadTest_Validator();
+		$v->setAllowedMaxFileSize($maxFileSizes);
+		
+		$retrievedSize = $v->getAllowedMaxFileSize('[doc]');
+		$this->assertEquals(2000, $retrievedSize, 'Max file size check on instance values failed');
+		
+		// Check that the instance values overwrote the default values
+		// ie. The max file size will not exist for [image]
+		$retrievedSize = $v->getAllowedMaxFileSize('[image]');
+		$this->assertFalse($retrievedSize, 'Max file size check on instance values failed');
```

Comment:
So the way `getAllowedMaxFileSize` and `setAllowedMaxFileSize` worked before was:
- If you set a new allowed max file size, it overrides anything that use to be set
- If nothing is set, return false

My code makes the allowed max file size have an initial value which is either what is defined in the configuration or what PHP says in the maximum upload/post size (if nothing is defined in the config). Because I am currently using the overall way it worked before, setting a new value override the old (rather than merge) which means `[image]` check actually returns false.

If I introduce the ability to merge the values, I should probably introduce a way to remove items and also clear the entire list. I can do this however my change was designed to be an addition without breaking change.

I do though understand what you mean, it would be nice if the values did merge rather than override. If you are happy with the scope of this PR to increase, I can introduce those changes too.